src/dmi_decode.c

```diff
@@ -0,0 +1,249 @@
+{
+    ssize_t r = 1;
+    size_t r2 = 0;
+
+    while (r2 != count && r != 0) {
+        r = read(fd, buf + r2, count - r2);
+        if (r == -1) {
+            if (errno != EINTR) {
+                perror(prefix);
+                return -1;
+            }
+        } else {
+            r2 += r;
+        }
+    }
+
+    if (r2 != count) {
+        fprintf(stderr, "%s: Unexpected end of file\n", prefix);
+        return -1;
+    }
+
+    return 0;
+}
+
+static void *read_file(off_t base, size_t *max_len, const char *filename)
+{
+    struct stat statbuf;
+    int fd;
+    uint8_t *p;
+
+    /*
+     * Don't print error message on missing file, as we will try to read
+     * files that may or may not be present.
+     */
+    if ((fd = open(filename, O_RDONLY)) == -1) {
+        if (errno != ENOENT)
+            perror(filename);
+
+        return NULL;
+    }
+
+    /*
+     * Check file size, don't allocate more than can be read.
+     */
+    if (fstat(fd, &statbuf) == 0) {
+        if (base >= statbuf.st_size) {
+            fprintf(stderr, "%s: Can't read data beyond EOF\n",
+                filename);
+            p = NULL;
+            goto out;
+        }
+        if (*max_len > (size_t)statbuf.st_size - base)
+            *max_len = statbuf.st_size - base;
+    }
+
+    if ((p = malloc(*max_len)) == NULL) {
+        perror("malloc");
+        goto out;
+    }
+
+    if (lseek(fd, base, SEEK_SET) == -1) {
+        fprintf(stderr, "%s: ", filename);
+        perror("lseek");
+        goto err_free;
+    }
+
+    if (att_read(fd, p, *max_len, filename) == 0)
+        goto out;
+
+err_free:
+    free(p);
+    p = NULL;
+
+out:
+    if (close(fd) == -1)
+        perror(filename);
+
+    return p;
+}
+
+static void ascii_filter(char *bp, size_t len)
+{
+    size_t i;
+
+    for (i = 0; i < len; i++)
+        if (bp[i] < 32 || bp[i] >= 127)
+            bp[i] = '.';
+}
+
+static char *_dmi_string(const struct dmi_header *dm, uint8_t s, int filter)
+{
+    char *bp = (char *)dm->data;
+  
+    bp += dm->length;
+    while (s > 1 && *bp) {
+        bp += strlen(bp);
+        bp++;
+        s--;
+    }
+
+    if (!*bp)
+        return NULL;
+
+    if (filter)
+        ascii_filter(bp, strlen(bp));
+
+    return bp;
+}
+
+const char *dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+    char *bp;
+  
+    if (s == 0)
+        return "Not Specified";
+  
+    bp = _dmi_string(dm, s, 1);
+    if (bp == NULL)
+          return bad_index;
+
+    return bp;
+}
+
+static int att_smbios3_decode(uint8_t *buf, const char *devmem, struct meta_data *meta)
+{
+    size_t len;
+    int ret = 0;
+    uint8_t *dmi_buf = NULL;
+
+    /* Don't let checksum run beyond the buffer */
+    if (buf[0x06] > 0x20) {
+        fprintf(stderr,
+            "Entry point length too large (%u bytes, expected %u).\n",
+            (unsigned int)buf[0x06], 0x18U);
+        return -1;
+    }
+
+    /* Some SMBIOS magic constants. I guess this is from specification. */
```

Comment:
Guess? Please don't use byte buffer if the structure is known (https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.4.0.pdf page 27).

Reply:
@krystian-hebel the `dmidecode` tool is the allocated buffer itself to iterate through it.